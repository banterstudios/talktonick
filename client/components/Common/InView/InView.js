import React, { PureComponent } from 'react'

import PropTypes from 'prop-types'

import { isElemInView, testForPassiveScroll } from 'client/utils/domUtils'

import { throttle } from 'client/utils/commonUtils'

class InView extends PureComponent {
  static propTypes = {
    onChange: PropTypes.func,
    scrollCheck: PropTypes.bool,
    resizeCheck: PropTypes.bool,
    timerCheck: PropTypes.bool,
    timerDelay: PropTypes.number
  }

  static defaultProps = {
    scrollCheck: true,
    resizeCheck: true,
    timerCheck: false,
    timerDelay: 200
  }

  constructor (props) {
    super(props)

    this.state = {
      isVisible: null
    }

    this.passiveData = testForPassiveScroll() ? { passive: true } : false
    this.containerRef = null
    this.timer = null
  }

  componentDidMount () {
    this.start()
  }

  componentWillUnmount () {
    this.unBindEvents()
  }

  start = () => {
    this.bindEvents()
    this.performCheck()
  }

  storeRef = (ref) => {
    this.containerRef = ref
  }

  bindEvents = () => {
    const { scrollCheck, resizeCheck, timerCheck, timerDelay } = this.props

    if (scrollCheck) {
      window.addEventListener('scroll', this.handleEvent, this.passiveData)
    }

    if (resizeCheck) {
      window.addEventListener('resize', this.handleEvent, false)
    }

    if (timerCheck) {
      this.timer = setInterval(this.performCheck, timerDelay)
    }
  }

  unBindEvents = () => {
    const { scrollCheck, resizeCheck, timerCheck } = this.props

    if (scrollCheck) {
      window.removeEventListener('scroll', this.handleEvent, this.passiveData)
    }

    if (resizeCheck) {
      window.removeEventListener('resize', this.handleEvent, false)
    }

    if (timerCheck) {
      this.timer = clearInterval(this.timer)
    }
  }

  handleEvent = throttle(this.performCheck, 100)

  performCheck = () => {
    const { onChange } = this.props

    const isVisible = isElemInView(this.containerRef)

    if (isVisible !== this.state.isVisible) {
      this.setState({ isVisible })
      onChange && onChange(isVisible)
    }
  }

  render () {
    const { children } = this.props

    return (
      <div ref={this.storeRef}>
        { children }
      </div>
    )
  }
}

export default InView
