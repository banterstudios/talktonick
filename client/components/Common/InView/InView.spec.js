import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import InView from './Inview'

const DummyComponent = () => <div />

jest.useFakeTimers()

describe('components/Common/InView', () => {
  let initialProps

  beforeEach(() => {
    initialProps = {
      onVisible: jest.fn(() => {
        console.log('yes!')
      }),
      onHidden: jest.fn(() => {
        console.log('no!')
      })
    }
  })

  describe('@renders', () => {
    it('in default state', () => {
      const component = shallow(<InView { ...initialProps } />)

      expect(toJson(component)).toMatchSnapshot()
      component.unmount()
    })

    it('with react node children', () => {
      const component = shallow(<InView { ...initialProps }><DummyComponent /></InView>)

      expect(toJson(component)).toMatchSnapshot()
      component.unmount()
    })

    it('with react element children', () => {
      const component = shallow(<InView { ...initialProps }><div /></InView>)

      expect(toJson(component)).toMatchSnapshot()
      component.unmount()
    })
  })

  describe('@instance methods', () => {
    describe('@handleEvent', () => {
      it('should call onVisible', () => {
        const component = shallow(<InView { ...initialProps } />)
        component.instance().handleEvent()

        expect(initialProps.onHidden).toHaveBeenCalledTimes(1)
        component.unmount()
      })
    })
  })
})
