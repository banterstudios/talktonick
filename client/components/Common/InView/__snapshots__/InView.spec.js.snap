// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Common/InView @renders in default state 1`] = `<div />`;

exports[`components/Common/InView @renders with react element children 1`] = `
<div>
  <div />
</div>
`;

exports[`components/Common/InView @renders with react node children 1`] = `
<div>
  <DummyComponent />
</div>
`;
